clients/src/main/java/org/apache/kafka/common/config/provider/EnvVarConfigProvider.java

```diff
@@ -0,0 +1,114 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.config.provider;
+
+import org.apache.kafka.common.config.ConfigData;
+import org.apache.kafka.common.config.ConfigException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.apache.kafka.common.config.provider.EnvVarConfigProviderConfig.ENV_VAR_CONFIG_PROVIDER_PATTERN_CONFIG;
+
+public class EnvVarConfigProvider implements ConfigProvider {
+    private final Map<String, String> envVarMap;
+    private Pattern envVarPattern;
+
+    public EnvVarConfigProvider() {
+        envVarMap = getEnvVars();
+    }
+
+    public EnvVarConfigProvider(Map<String, String> envVarsAsArgument) {
+        envVarMap = envVarsAsArgument;
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(EnvVarConfigProvider.class);
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        if (configs.keySet().contains(ENV_VAR_CONFIG_PROVIDER_PATTERN_CONFIG)) {
+            envVarPattern = Pattern.compile(
+                    String.valueOf(configs.get(ENV_VAR_CONFIG_PROVIDER_PATTERN_CONFIG))
+            );
+        } else {
+            envVarPattern = Pattern.compile(".*");
+            log.info("No pattern for environment variables provided. Using default pattern '(.*)'.");
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+    }
+
+    /**
+     * @param s unused
+     * @return returns environment variables as configuration
+     */
+    @Override
+    public ConfigData get(String s) {
+        return get(s, null);
+    }
+
+    /**
+     * @param path    path, not used for environment variables
+     * @param keys the keys whose values will be retrieved.
+     * @return the configuration data.
+     */
+    @Override
+    public ConfigData get(String path, Set<String> keys) {
+
+        if (path != null && !path.isEmpty()) {
+            log.error("Path is not supported for EnvVarConfigProvider, invalid value '{}'", path);
+            throw new ConfigException("Path is not supported for EnvVarConfigProvider, invalid value '" + path + "'");
+        }
+
+        if (envVarMap == null) {
+            return new ConfigData(new HashMap<>());
+        }
+
+        Map<String, String> filteredEnvVarMap = envVarMap;
+
+        filteredEnvVarMap = envVarMap.entrySet().stream()
```

Comment:
Nit: can be simplified

```suggestion
        Map<String, String> filteredEnvVarMap = envVarMap.entrySet().stream()
```